src/main/java/org/testng/internal/reflect/DataProviderMethodMatcher.java

```diff
@@ -0,0 +1,67 @@
+package org.testng.internal.reflect;
+
+import org.testng.ITestContext;
+import org.testng.ITestResult;
+
+import java.lang.reflect.Method;
+
+/**
+ * Checks the conformance as per data-provide specifications.
+ *
+ * @author <a href="mailto:nitin.matrix@gmail.com">Nitin Verma</a>
+ */
+public class DataProviderMethodMatcher implements MethodMatcher {
```

Comment:
Following examples are to depict responsibilities of each of the `MethodMatcher`. 
Assume `foo(A,B,C)` is a function and `A,B,C` are types. consider lowercase entities as type's instances.
Example1:
Method: `foo(A,B,C)`
Arguments: `{a,b,c}`
1. DataProviderMethodMatcher would check for a direct match.
2. As its a direct match DirectMethodMatcher would confirm and returns `{a,b,c}`.
3. Method `foo` would be reflectively invoked with args `{a,b,c}`

Example2:
Method: `foo(A,B,C...)`
Arguments: `{a,b,c1,c2}`

1. DataProviderMethodMatcher would check for a direct match.
2. Its not a direct match and DirectMethodMatcher would indicate so to the caller.
3. Now DataProviderMethodMatcher would check for an array ending match.
4. ArrayEndingMethodMatcher looking at ending array in parameters matches ending types to C and check if input args can fit `{a,b,{c1,c2}}` formation. confirms and returns `{a,b,{c1,c2}}` .
5. Method `foo` would be reflectively invoked with args `{a,b,{c1,c2}}`

Example2:
Method: `foo(A,B,C...)`
Arguments:` {a,b,{c1,c2,c3}}`

1. DataProviderMethodMatcher would check for a direct match.
2. As its a direct match DirectMethodMatcher would confirm and return `{a,b,{c1,c2,c3}}`.
3. Method `foo` would be reflectively invoked with args `{a,b,{c1,c2,c3}}`